dask/dataframe/shuffle.py

```diff
@@ -24,6 +26,296 @@
 logger = logging.getLogger(__name__)
 
 
+class SimpleShuffleLayer(Layer):
+    """Simple HighLevelGraph Shuffle layer
+
+    High-level graph layer for a simple shuffle operation in which
+    each output partition depends on all input partitions.
+
+    Parameters
+    ----------
+    name : str
+        Name of new shuffled output collection.
+    column : str or list of str
+        Column(s) to be used to map rows to output partitions (by hashing).
+    npartitions : int
+        Number of output partitions.
+    npartitions_input : int
+        Number of partitions in the original (un-shuffled) DataFrame.
+    ignore_index: bool, default False
+        Ignore index during shuffle.  If ``True``, performance may improve,
+        but index values will not be preserved.
+    dep_name : str
+        Name of input collection.
+    dep_meta : pd.DataFrame-like object
+        Empty metadata of input collection.
+    parts_out : list of int (optional)
+        List of required output-partition indices.
+    """
+
+    def __init__(
+        self,
+        name,
+        column,
+        npartitions,
+        npartitions_input,
+        ignore_index,
+        dep_name,
+        dep_meta,
+        parts_out=None,
+    ):
+        self.name = name
+        self.column = column
+        self.npartitions = npartitions
+        self.npartitions_input = npartitions_input
+        self.ignore_index = ignore_index
+        self.dep_name = dep_name
+        self.dep_meta = dep_meta
+        self.parts_out = parts_out or range(npartitions)
+
+    def __repr__(self):
+        return "SimpleShuffleLayer<name='{}', npartitions={}>".format(
+            self.name, self.npartitions
+        )
+
+    @property
+    def _dict(self):
+        """Materialize full dict representation"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict
+        else:
+            dsk = self._construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
+
+    def _keys_to_parts(self, keys):
+        """Simple utility to convert keys to partition indices."""
+        parts = set()
+        for key in keys:
+            try:
+                _name, _part = key
+            except ValueError:
+                continue
+            if _name != self.name:
+                continue
+            parts.add(_part)
+        return parts
+
+    def _cull_dependencies(self, keys, parts_out=None):
+        """Determine the necessary dependencies to produce `keys`.
+
+        For a simple shuffle, output partitions always depend on
+        all input partitions. This method does not require graph
+        materialization.
+        """
+        deps = defaultdict(set)
```

Comment:
@madsbk - I just wanted to get your thoughts on something here...

It seems that the HLG `cull` operation breaks when this `deps` dictionary is a vanilla `dict` (rather than a `defaultdict(set)`).  The problem is that we are not actually returning any intra-layer dependencies when the shuffle layer is culled (because we are not actually materializing the graph). This means, we will get a keyerror during this [culled_deps[k]](https://github.com/dask/dask/blob/51b76e5d4820c435c4ee179ef31ea9916ec229c8/dask/highlevelgraph.py#L435) access in `HighLevelGraph.cull` (because we are not adding dependencies for **all** keys).

Does this seem like it could be a problem?

Reply:
This should be fixed in https://github.com/dask/dask/pull/6699, in which `cull()` is only required to return _external_ dependencies.

Reply:
Thanks @madsbk !